### PR TITLE
Add an end to end test for the server.

### PR DESCRIPTION
**Pull request by melaniedc**
This sets up some basic cases for interactions between the backend and the server. It will definitely need more filling in, but this gives us a basic framework to build off of.

### CODE REVIEW

File: tests/test_end_to_end.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        server.app.config["VariantBackend"] = backend
+        self.app = server.app.test_client()
+
+    def sendJSONPostRequest(self, path, data):
+        return self.app.post(path,
+                             headers={'Content-type': 'application/json'},
+                             data=data)
+
+    def runVariantSetsSearch(self):
+        expectedIds = ['example_1', 'example_2', 'example_3', 'example_4']
```
Comment (jeromekelleher) [inline tests/test_end_to_end.py]: This is fine for now, but we should modify our test backend API to return the variant sets we want so we don't need to hard code these values. Could you insert a TODO comment to this effect here please?

File: tests/test_end_to_end.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        self.assertTrue(protocol.GASearchVariantSetsResponse.validate(
+            responseData.toJSONDict()))
+
+        self.assertIsNone(responseData.nextPageToken)
+        self.assertEqual(4, len(responseData.variantSets))
+        for variantSet in responseData.variantSets:
+            self.assertTrue(variantSet.id in expectedIds)
+
+    def runVariantsSearch(self):
+        expectedIds = ['example_1']
```
Comment (jeromekelleher) [inline tests/test_end_to_end.py]: Please insert a line `referenceName = '1'` and replace the hard-coded values below. Also add a TODO comment stating we should get these values from the test backend when it's being refactored.

File: tests/test_end_to_end.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+        self.assertEqual(200, response.status_code)
+
+        responseData = protocol.GASearchVariantSetsResponse.fromJSON(
+            response.data)
+        self.assertTrue(protocol.GASearchVariantSetsResponse.validate(
+            responseData.toJSONDict()))
+
+        self.assertIsNone(responseData.nextPageToken)
+        self.assertEqual(4, len(responseData.variantSets))
```
Comment (jeromekelleher) [inline tests/test_end_to_end.py]: It would be better to use `len(expectedIds)` here rather than 4.

**Comment by jeromekelleher** [comment]
This is excellent, thank @melaniedc! We will need to generalise this to work against unknown test data in the future, and I've added a few requests to make this explicit. There's nothing particularly special about the test data we're using at the moment, and I'm certain we'll have to add more as bugs arise that our current test set won't provoke. 

We're going to need a big refactor of the test system at some point, but I'd rather wait to see what other great stuff is going to land before doing this.

**Comment by melaniedc** [comment]
Agreed. My main goal here is to set up the basic framework so someone with a better grasp of the biology/functionality can fill in more useful scenarios :)

**Comment by melaniedc** [comment]
This now works with the new backend organization :)

**Comment by kerrydc** [comment]
+1, looks good.

**Comment by jeromekelleher** [comment]
Looking at this again, I wonder if perhaps we should reorganise it a little to make the test cases more modular. As it, we only get one test case from this module, `testWormtable`. As we add more and more tests, this one test gets more and more complicated, but we don't get any more test cases (from unittest's perspective). It would be nice to change `runVariantSetsSearch` to `testVariantSetsSearch` and do the same for variants. This way they are tested independently, and failing one test doesn't prevent the execution of the other. Ideally, we'd like to handle the fixtures in the `setUp` and `tearDown` methods of the test class, but since `WormtableTestFixture` is quite heavy we should probably handle it in the module `setUp` and `tearDown` as is done in `test_backends`. Another advantage of this is that `tearDown` is guaranteed to happen --- at the moment, if an error occurs, the WormtableTestFixture will not be cleaned up, leaving a mess behind in /tmp.

**Comment by melaniedc** [comment]
Sounds good - especially with the way the backends stuff is evolving I think that makes more sense. We'll need to re-organize these some more once we work out #82 and the tabix backends tests but for now I'm just making it wormtable specific since that's what's implemented :)

File: tests/test_end_to_end.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
+"""
+End-to-end tests for each backend configuration. Sets up a server with
+the backend, sends some basic queries to that server and verifies results
+are as expected.
+"""
```
Comment (jeromekelleher) [inline tests/test_end_to_end.py]: Can we get the __future__ imports here please? These are important for py2 -> py3 compatibility, and we definitely want to support py3 at some point!

**Comment by jeromekelleher** [comment]
Looks good to me! I'm +1 to merge, modulo the future imports.

**Comment by melaniedc** [comment]
Done and squashed :)

**Comment by dcolligan** [comment]
+1

**Comment by Naburimannu** [comment]
Are we mixing end-to-end tests with unit tests? I like keeping unit tests fast and separating out other things like integration and benchmarking.

**Comment by jeromekelleher** [comment]
I see this as a quick unit test, just to make sure everything is plumbing together correctly. It's a bit slow now because of the test fixtures required, but hopefully we can fix this soon. What bit are you concerned about here @Naburimannu?

**Comment by Naburimannu** [comment]
Sorry I left this idle. +1; I'll file an issue to discuss separating unit, integration, and end-to-end tests. @melaniedc pointed out that we want to be running at least a few end-to-end tests in Travis, and for now having all the tests in one bucket is what we've got.

**Comment by jeromekelleher** [comment]
Merging, thank @melaniedc!

**Pull request merged**